src/Microsoft.Extensions.OData.Migration/UriSegmentTranslator.cs

```diff
@@ -0,0 +1,243 @@
+            string v3typeName = segment.Property.DeclaringType.FullTypeName();
+
+            // Look up V4 Structured Type (guaranteed if found) in V4 model
+            IEdmStructuredType v4Type = v4model.FindType(v3typeName) as IEdmStructuredType;
+
+            // Look up corresponding Property in V4 Type by name
+            IEdmStructuralProperty v4Property = v4Type.FindProperty(segment.Property.Name) as IEdmStructuralProperty;
+
+            return new PropertySegment(v4Property);
+        }
+
+        /// <summary>
+        /// Translates a V3 OperationSegment to V4 OperationSegment by locating all matching function imports in the v4 model
+        /// and constructing a new V4 OperationSegment with all found operation segments.
+        /// </summary>
+        /// <param name="segment"></param>
+        /// <returns></returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.OperationSegment segment)
+        {
+            List<IEdmOperationImport> v4OpImports = new List<IEdmOperationImport>();
+            List<IEdmOperation> v4UnboundOps = new List<IEdmOperation>();
+            List<IEdmOperation> v4BoundOps = new List<IEdmOperation>();
+
+            IEdmEntitySet v4entitySet = segment.EntitySet != null ? v4model.FindDeclaredEntitySet(segment.EntitySet.Name) : null;
+
+            // For each operation in V3 operation segment, determine corresponding action/function/function imports in v4
+            foreach (Data.Edm.IEdmFunctionImport op in segment.Operations)
+            {
+                IEnumerable<Data.Edm.IEdmFunctionParameter> entityParams = op.Parameters.Where(p => p.Type.Definition.GetType().GetInterfaces().Select(iface => iface.Name).Contains("IEdmEntityType"));
+                foreach (var p in entityParams)
+                {
+                    // Find bounded operations
+                    IEnumerable<IEdmOperation> boundOps = v4model.FindBoundOperations(v4model.FindType(p.Type.Definition.FullTypeName()));
+                    v4BoundOps.AddRange(boundOps.Where(boundOp => boundOp.FullName() == op.Container.Namespace + "." + op.Name));
+                }
+
+                // Should I be searching also by container namespace and name?
+                string searchQuery = op.Name;
+
+                // Find v4 operation imports
+                IEnumerable<IEdmOperationImport> foundOperationImports = v4model.FindDeclaredOperationImports(searchQuery);
+                v4OpImports.AddRange(foundOperationImports);
+
+                // Find v4 unbounded operations
+                IEnumerable<IEdmOperation> foundOperations = v4model.FindOperations(searchQuery);
+                v4UnboundOps.AddRange(foundOperations);
+            }
+
+            // If function imports are found, return OperationImportSegment
+            if (v4OpImports.Count() > 0)
+            {
+                return new OperationImportSegment(v4OpImports.First(), v4entitySet);

```

Comment:
Could v4OpImports contain multiple items? If yes, is it fine to take the firs one only?

Reply:
Not sure, will be more clear with examples